### PR TITLE
Teams 1

### LINKED ISSUE #71: Migrate team

### PR DESCRIPTION
**Pull request by Josh-Dykstra**
fixes #71

### CODE REVIEW

File: packages/plugin-conversation/src/conversation.js [hunk 1]
```diff
@@ -561,7 +561,7 @@ const Conversation = SparkPlugin.extend({
       kmsMessage: {
         method: `create`,
         uri: `/resources`,
-        userIds: params.participants,
+        userIds: cloneDeep(params.participants),
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/conversation.js]: Before this change userIds held a reference to params.participants. This mean that any changes in params.participants was also reflected here leading so some unexpected side effects. With this change kms.userIds will have its own array of user ids

File: packages/plugin-team/README.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
+# addConversation
```
Comment (Josh-Dykstra) [inline packages/plugin-team/README.md]: Should this file be committed?

File: packages/plugin-team/src/team.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  get(team, options) {
+    options = options || {};
+
+    if (!team.id) {
+      return Promise.reject(new Error(`'team.id' is required`));
+    }
+
+    return this.request({
+      api: `conversation`,
+      resource: `teams/${team.id}`,
```
Comment (Josh-Dykstra) [inline packages/plugin-team/src/team.js]: Would it be better to use the `team.url` instead of synthesizing it from the team's id?

File: packages/plugin-team/README.md (hunk 2 above)
Comment (ianwremmel) [inline packages/plugin-team/README.md]: Yea, it'll be overridden in the build/test/publish process

File: packages/plugin-team/src/decrypter.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  if (usableKey && team.displayName) {
+    try {
+      const parts = team.displayName.split(`.`);
+      if (parts.length === 5) {
+        JSON.parse(base64.decode(parts[0]));
+
+        // Much like a conversation, a team can be the object of update, add or leave activities.
+        // If this is the case we measure decryption of parent object
+        this.spark.encryption.metrics.startDecryptionMetrics(parentObject || team);
```
Comment (ianwremmel) [inline packages/plugin-team/src/decrypter.js]: encryption metrics doesn't exist at the moment; feel free to remove the metrics lines in here.

File: packages/plugin-team/src/decrypter.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+Decrypter.prototype._decryptTeam = function _decryptTeam(team, key, options, parentObject) {
+  let promises = [];
+  if (team.conversations.items) {
+    promises = team.conversations.items.map((item) => this.decryptObject(item, options));
+  }
+
+  const usableKey = key || team.encryptionKeyUrl;
+
+  if (usableKey && team.displayName) {
+    try {
```
Comment (ianwremmel) [inline packages/plugin-team/src/decrypter.js]: This try/catch was inspired by some code from elsewhere, but I can't find it anymore. Any idea where this concept came from?

File: packages/plugin-team/src/decrypter.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        JSON.parse(base64.decode(parts[0]));
+
+        // Much like a conversation, a team can be the object of update, add or leave activities.
+        // If this is the case we measure decryption of parent object
+        this.spark.encryption.metrics.startDecryptionMetrics(parentObject || team);
+
+        promises.push(this.decryptProperty(team, `displayName`, usableKey, () => {
+          team.displayName = this.spark.encryption.config.decryptionFailureMessage;
+        }));
+        promises.push(this.decryptProperty(team, `summary`, usableKey, () => {
```
Comment (ianwremmel) [inline packages/plugin-team/src/decrypter.js]: decryptProperty no longer accepts the failure handler (in large part because the it never really seemed to work that well). Let's consider doing something more generic; maybe we can add an object of undecryptableKeyPaths to the thing that can't be decrypted and pass the error message up to the client.

File: packages/plugin-team/src/encrypter.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+/**
+ * Encrypts a summary
+ * @param {string} summary
+ * @param {Encryption~Key|string} key
+ * @private
+ * @returns {Promise} Resolves with the encrypted string
+ */
+Encrypter.prototype._encryptPropSummary = function _encryptPropSummary(summary, key) {
```
Comment (ianwremmel) [inline packages/plugin-team/src/encrypter.js]: Consider setting all these new functions using Object.assign, that way we'll get the collision handling that we don't have in e.g. SquaredAdapter.js

File: packages/plugin-team/src/team.js [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+  namespace: `Team`,
+
+  /**
+   * Move an existing group conversation into a team.
+   * @param {TeamObject} team
+   * @param {ConversationObject} conversation
+   * @returns {Promise} Resolves with the add activity
+   */
+  addConversation(team, conversation) {
+
```
Comment (ianwremmel) [inline packages/plugin-team/src/team.js]: Did you forget to push or is this a work in progress PR?

File: packages/plugin-team/src/team.js [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+  /**
+   * Create a team
+   * @param {NewTeamObject} params
+   * @returns {Promise} Resolves with the created team
+   */
+  create(params) {
+
+    if (!params.displayName) {
+      return Promise.reject(new Error(`'params.displayName' is required`));
```
Comment (ianwremmel) [inline packages/plugin-team/src/team.js]: use escaped backticks instead of apostrophes

File: packages/plugin-team/src/team.js [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+      .then((res) => this._recordUUIDs(res.body)
+        .then(() => res.body));
+  },
+
+  /**
+   * Get the list of conversations for a particular team
+   * @param {TeamObject} team
+   * @returns {Promise} Resolves with an array of conversations
+   */
+  getConversations(team) {
```
Comment (ianwremmel) [inline packages/plugin-team/src/team.js]: listConversations

File: packages/plugin-team/src/team.js (hunk 3 above)
Comment (ianwremmel) [inline packages/plugin-team/src/team.js]: yes

File: packages/plugin-team/src/team.js (hunk 10 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/team.js]: Good point.

File: packages/plugin-team/src/team.js (hunk 9 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/team.js]: Sure.

File: packages/plugin-team/src/team.js (hunk 8 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/team.js]: I just added the stubs for all the action methods. They will be implemented in Teams 2 PR

**Comment by Josh-Dykstra** [comment]
Yeah planning on having this PR be team creation and retrival. Then another with verbs.

File: packages/plugin-team/src/team.js (hunk 3 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/team.js]: Done.

File: packages/plugin-team/src/decrypter.js (hunk 5 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/decrypter.js]: The original conversation [decrypter](https://github.com/ciscospark/spark-js-sdk/blob/master/src/client/services/conversation/decrypter.js#L164-L192) was really defensive, so the teams one was as well. Will change this to be more like the new and improved plugin-conversation encrypter.

File: packages/plugin-team/src/decrypter.js (hunk 4 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/decrypter.js]: Got it. Thanks.

File: packages/plugin-team/src/encrypter.js (hunk 7 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/encrypter.js]: See amended commit.

File: packages/plugin-team/src/encrypter.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
+/**!
+ *
+ * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
+ * @private
+ */
+
+import {Encrypter} from '@ciscospark/plugin-conversation';
+import {isArray} from 'lodash';
+
+Object.assign(Encrypter, {
```
Comment (ianwremmel) [inline packages/plugin-team/src/encrypter.js]: Encrypter.prototype

File: packages/plugin-team/src/decrypter.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
+/**!
+ *
+ * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
+ * @private
+ */
+import {Decrypter} from '@ciscospark/plugin-conversation';
+
+Object.assign(Decrypter, {
```
Comment (ianwremmel) [inline packages/plugin-team/src/decrypter.js]: Decrypter.prototype

File: packages/plugin-team/src/team.js [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+   * @param {string} team.id
+   * @param {Object} options
+   * @param {boolean} options.includeTeamConversations
+   * @param {boolean} options.includeTeamMembers
+   * @returns {Promise} Resolves with the requested team
+   */
+  get(team, options) {
+    options = options || {};
+
+    return this._inferTeamUrl(team)
```
Comment (ianwremmel) [inline packages/plugin-team/src/team.js]: Take a look at `Conversation#list()` and `Conversation#get()`. I changed to it be much more accepting of parameters and to default to requesting as little data as possible.

File: packages/plugin-team/test/integration/spec/actions.js [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  describe(`#assignModerator()`, () => {
+    it(`assigns a team moderator`, () => {});
+  });
+
+  describe(`#archive()`, () => {
+    it(`archives a team`, () => {});
+    it(`archives a team conversation`, () => {});
+  });
+
+  describe(`#joinConversation()`, () => {
```
Comment (ianwremmel) [inline packages/plugin-team/test/integration/spec/actions.js]: Why is this necessary? Does it make sense to make this part of conversation?

File: packages/plugin-team/test/integration/spec/create.js [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+        displayName,
+        participants: [kirk, spock]
+      })
+        .then((tc) => {
+          assert.isInternalTeamConversation(tc);
+          assert.lengthOf(tc.participants.items, 2);
+        });
+    });
+
+    it(`creates a team using the 'includeAllTeamMembers' parameter`, () => {
```
Comment (ianwremmel) [inline packages/plugin-team/test/integration/spec/create.js]: Maybe change this description so its clear what that parameter does?

File: packages/plugin-team/src/encrypter.js (hunk 11 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/encrypter.js]: Yeah meant to do this. Good catch.

File: packages/plugin-team/test/integration/spec/actions.js (hunk 14 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/test/integration/spec/actions.js]: This is the function used to add yourself to an open (team) conversation. This is a feature that only exists for team conversations and not of typical group conversations so in some ways it seems like it should be in the teams module... When implemented it'll be putting a POST request on the `teams` resource, too.  I see what you're saying, though... could fit in either place if you think about it.

File: packages/plugin-team/test/integration/spec/create.js (hunk 15 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/test/integration/spec/create.js]: Good call.

File: packages/plugin-team/src/team.js (hunk 13 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/team.js]: I see. We used to whitelist params and now we're accepting all params. Seems like a good choice - now we won't need a change whenever a new param is added to the service. Any reason in particular we essentially blacklist these ones here? https://github.com/ciscospark/spark-js-sdk/blob/master/packages/plugin-conversation/src/conversation.js#L206

File: packages/plugin-conversation/src/decrypter.js [hunk 16]
```diff
@@ -59,6 +59,12 @@ const Decrypter = SparkPlugin.extend({
 
     return this[methodName](key, object[property])
       .then((decryptedProperty) => {
+
+        if (this.config.conversation.keepEncryptedProperties) {
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/decrypter.js]: Bringing back this config option so that we can compare values in tests

File: packages/plugin-conversation/src/index.js [hunk 17]
```diff
@@ -21,3 +21,6 @@ registerPlugin(`conversation`, Conversation, {
 
 export {default as default} from './conversation';
 export {default as ShareActivity} from './share-activity';
+export {default as Encrypter} from './encrypter';
+export {default as Decrypter} from './decrypter';
+export {default as Normalizer} from './normalizer';
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/index.js]: Exposing these so they can be extended by teams module.

File: packages/plugin-conversation/src/interceptor.js [hunk 18]
```diff
@@ -60,9 +60,8 @@ export default class ConversationInterceptor extends Interceptor {
         }
 
         const hasItems = Boolean(response.body.items);
-
-        return this.spark.conversation.decrypter.decryptObject(null, response.body)
-          .then((body) => this.spark.conversation.normalizer.normalize(hasItems ? body.items : body))
+        return this.spark.conversation.decrypter.decryptObject(null, hasItems ? response.body.items : response.body)
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/interceptor.js]: If the response has an items attribute, Decrypter needs the items and not `{items: [...]}`

File: packages/plugin-team/src/normalizer.js [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+ * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
+ * @private
+ */
+
+import Normalizer from '@ciscospark/plugin-conversation';
+
+
+Object.assign(Normalizer.prototype, {
+
+  normalizeTeam(team) {
```
Comment (Josh-Dykstra) [inline packages/plugin-team/src/normalizer.js]: Need docs

File: packages/plugin-team/src/index.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+import Team from './team';
+import config from './config';
+
+import '@ciscospark/plugin-conversation';
+import '@ciscospark/plugin-user';
+import '@ciscospark/plugin-encryption';
+
+import './encrypter';
+import './decrypter';
+import './normalizer';
```
Comment (Josh-Dykstra) [inline packages/plugin-team/src/index.js]: Pull these in so that they will extend their conversation counterparts

File: packages/plugin-team/src/team.js (hunk 8 above)
Comment (Josh-Dykstra) [inline packages/plugin-team/src/team.js]: This is how you add an existing group room to a team

File: packages/plugin-team/test/integration/spec/create.js [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+      });
+
+      return Promise.all([
+        spark.mercury.connect(),
+        spark.device.register()
+      ]);
+    })
+  );
+
+  after(() => spark.mercury.disconnect());
```
Comment (Josh-Dykstra) [inline packages/plugin-team/test/integration/spec/create.js]: Is there any device cleanup we need?

File: packages/plugin-conversation/src/interceptor.js [hunk 22]
```diff
@@ -34,7 +34,7 @@ export default class ConversationInterceptor extends Interceptor {
           return options;
         }
 
-        if (options.resource !== `content` && options.resource !== `activities` && options.resource !== `conversations`) {
+        if (options.resource !== `content` && options.resource !== `activities` && options.resource !== `conversations` && options.resource !== `teams`) {
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/interceptor.js]: May need to open an issue for a more clever way to do this.

File: packages/plugin-conversation/src/interceptor.js (hunk 18 above)
Comment (ianwremmel) [inline packages/plugin-conversation/src/interceptor.js]: i feel like i've fixed this a few times...too many outstanding PRs right now...

File: packages/plugin-team/README.md (hunk 2 above)
Comment (ianwremmel) [inline packages/plugin-team/README.md]: Follow up: it should be committed, but empty. with the no-autogen-readmes PR, you'll have some unfortunate conflicts. If I get that one merged before this gets merged, change this readme to be simple description and a pointer to the docs site.

File: packages/plugin-team/test/integration/spec/actions.js (hunk 14 above)
Comment (ianwremmel) [inline packages/plugin-team/test/integration/spec/actions.js]: Gotcha. I feel like I asked this question a year ago...

File: packages/plugin-team/test/integration/spec/create.js (hunk 21 above)
Comment (ianwremmel) [inline packages/plugin-team/test/integration/spec/create.js]: nope

File: packages/plugin-team/test/integration/spec/get.js [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+        config: {
+          conversation: {
+            keepEncryptedProperties: true
+          }
+        }
+      });
+
+      return Promise.all([
+        kirk.spark.mercury.connect(),
+        kirk.spark.device.register(),
```
Comment (ianwremmel) [inline packages/plugin-team/test/integration/spec/get.js]: No need to call register, it's implicit from connect.

File: packages/test-helper-chai/src/assertions/schemas.js [hunk 24]
```diff
@@ -48,6 +48,13 @@ module.exports = function schemas(chai) {
     assert.property(this._obj, 'url');
   });
 
+  Assertion.addProperty('InternalTeam', function() {
```
Comment (ianwremmel) [inline packages/test-helper-chai/src/assertions/schemas.js]: Can you add the encryptedProperties as appropriate to these assertions?

File: packages/plugin-team/test/integration/spec/get.js [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+          kirk.spark.team.createConversation(team0, emptyRoom)
+        ]);
+      })
+      .then((conversations) => {
+        [teamConvo0, teamConvo1] = conversations;
+      });
+  });
+
+  after(() => Promise.all([
+    kirk.spark.mercury.disconnect(),
```
Comment (ianwremmel) [inline packages/plugin-team/test/integration/spec/get.js]: something to add in the next PR: make sure to check for existence of kirk and spock. if the before() fails becuase CI, the after will add an extra line to the test report (because kirk and spock are undefined).

in other words, change this to 

after(() => Promise.all([
  kirk && kirk.spark.mercury.disconnect(),
  spock && spock.spark.mercury.disconnect()
]);

Comment (Josh-Dykstra) [inline packages/plugin-team/test/integration/spec/get.js]: Will add this check. Thanks.
